src/librustc_trans/mir/place.rs

```diff
@@ -266,7 +266,13 @@ impl<'a, 'tcx> PlaceRef<'tcx> {
         let cast_to = bx.cx.layout_of(cast_to).immediate_llvm_type(bx.cx);
         match self.layout.variants {
             layout::Variants::Single { index } => {
-                return C_uint(cast_to, index as u64);
+                if self.layout.for_variant(bx.cx, index).abi == layout::Abi::Uninhabited {
+                    return C_undef(cast_to);
+                }
+                let discr_val = self.layout.ty.ty_adt_def().map_or(
+                    index as u64,
+                    |def| def.discriminant_for_variant(bx.cx.tcx, index).val as u64);
```

Comment:
You should use `C_big_uint` to support `u128`.

Reply:
Done.